### PR TITLE
Optimize using Span<T> Assert.Equal for T[] where T is unmanaged

### LINKED ISSUE #1907: Assert.Equals on two large arrays takes a long time
I noticed this when comparing 6GB worth of array data.  Assert.Equals takes a very long time: over 10 minutes.  Using Span.SequenceEqual instead is much faster: under 10s.  Here's a micro comparison on a 1MB array that shows a 220x difference.  Sample code attached. [xunitPerf.zip](https://github.com/xunit/xunit/files/2917368/xunitPerf.zip)

Method | Elapsed
------------ | -------------
Assert.Equal | 00:00:00.3534817
Span.SequenceEqual | 00:00:00.0015998

Not sure how many folks are doing assertions with large arrays, but you have an opportunity for an overload with better performance.

### PR DESCRIPTION
**Pull request by louis-z**
Fixes https://github.com/xunit/xunit/issues/1907

### CODE REVIEW

File: EqualityAsserts.cs [hunk 1]
```diff
@@ -9,6 +9,10 @@
 using System.Collections.Generic;
 using System.Globalization;
 using Xunit.Sdk;
+#if XUNIT_ASSERT_ARRAY_AS_SPAN
```
Comment (bradwilson) [inline EqualityAsserts.cs]: This is not necessary. Based on https://www.nuget.org/packages/system.memory it appears that Span can be made functional everywhere that we support assertions.

File: EqualityAsserts.cs [hunk 2]
```diff
@@ -9,6 +9,10 @@
...
+using System.Linq;
+using static System.MemoryExtensions;
```
Comment (bradwilson) [inline EqualityAsserts.cs]: I couldn't find any reason for this line. The code compiles without it.

File: EqualityAsserts.cs [hunk 3]
```diff
@@ -19,6 +23,33 @@ namespace Xunit
...
+			where T : unmanaged, IEquatable<T>
+		{
+			if (expected == null && actual == null)
+				return;
+			if (expected == null || actual == null)
+				throw new EqualException(expected, actual);
+			if (!expected.AsSpan().SequenceEqual(actual))
+				Assert.Equal((object)expected, (object)actual);
+		}
+#endif
```
Comment (bradwilson) [inline EqualityAsserts.cs:48]: nit: missing blank line between methods.

File: EqualityAsserts.cs [hunk 4]
```diff
@@ -170,7 +201,29 @@ public static void StrictEqual<T>(T expected, T actual)
 		{
 			Equal(expected, actual, EqualityComparer<T>.Default);
 		}
-
+#if XUNIT_ASSERT_ARRAY_AS_SPAN
```
Comment (bradwilson) [inline EqualityAsserts.cs]: nit: missing blank line between methods.

File: EqualityAsserts.cs [hunk 5]
```diff
@@ -170,7 +201,29 @@ public static void StrictEqual<T>(T expected, T actual)
...
+			where T : unmanaged, IEquatable<T>
+		{
+			if (expected == null && actual == null)
+				throw new NotEqualException(ArgumentFormatter.Format(expected), ArgumentFormatter.Format(actual));
+			if (expected == null || actual == null)
+				return;
+			if (expected.AsSpan().SequenceEqual(actual))
+				throw new NotEqualException(ArgumentFormatter.Format(expected), ArgumentFormatter.Format(actual));
+		}
+#endif
```
Comment (bradwilson) [inline EqualityAsserts.cs:225]: nit: missing blank line between methods.

**Review by bradwilson — CHANGES_REQUESTED** [review]
There are a few changes I'd like to see here, please see the comments inline.

File: EqualityAsserts.cs (hunk 1 above)
Comment (bradwilson) [reply EqualityAsserts.cs]: Actually, I take it back. There are things here that require C# 7.3 so we do still need this. That said, there will be more Span-related things coming, so let's change this to `XUNIT_SPAN`, and please move it up to the top as a separate block, like you see I've done for `XUNIT_NULLABLE`.

File: EqualityAsserts.cs [hunk 6]
```diff
@@ -170,7 +201,29 @@ public static void StrictEqual<T>(T expected, T actual)
...
+		public static void NotEqual<T>(T[] expected, T[] actual)
+#endif
+			where T : unmanaged, IEquatable<T>
+		{
+			if (expected == null && actual == null)
+				throw new NotEqualException(ArgumentFormatter.Format(expected), ArgumentFormatter.Format(actual));
+			if (expected == null || actual == null)
+				return;
+			if (expected.AsSpan().SequenceEqual(actual))
+				throw new NotEqualException(ArgumentFormatter.Format(expected), ArgumentFormatter.Format(actual));
```
Comment (bradwilson) [inline EqualityAsserts.cs:224]: This needs to be changed to Assert.NotEqual, like was done for Assert.Equal above.

File: EqualityAsserts.cs (hunk 1 above)
Comment (louis-z) [reply EqualityAsserts.cs]: It turns out I don't need that block after all.

File: EqualityAsserts.cs (hunk 2 above)
Comment (louis-z) [reply EqualityAsserts.cs]: You're absolutely right. Sorry about that.

**Comment by bradwilson** [comment]
Thanks!

**Comment by louis-z** [comment]
Glad I could help.

**Pull request merged**